### PR TITLE
[Translator] Set sources when extracting strings from php files

### PR DESCRIPTION
**Pull request by Stadly**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

Set metadata about filenames and line numbers when extracting translatable strings from php files. This PR complements #30909 and #31248, as those PRs implement support for dumping the filenames and line numbers to Po and Qt file formats.

### CODE REVIEW

**Comment by Stadly** [comment]
I don't understand why the test is failing on PHP7.3 I've tested with PHP7.3 locally, and it works fine. Can it have something to do with `deps=low`?

File: src/Symfony/Component/Translation/Extractor/PhpExtractor.php [hunk 1]
```diff
@@ -196,9 +196,10 @@ private function getValue(\Iterator $tokenIterator)
      * Extracts trans message from PHP tokens.
      *
      * @param array            $tokens
+     * @param string           $filename
      * @param MessageCatalogue $catalog
      */
-    protected function parseTokens($tokens, MessageCatalogue $catalog)
+    protected function parseTokens($tokens, $filename, MessageCatalogue $catalog)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Translation/Extractor/PhpExtractor.php]: That's a BC - it's not allowed.
I'd suggest adding the filename as last argument, but commented for now.
See how it's done e.g. on https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DomCrawler/Crawler.php#L502

Comment (Stadly) [reply src/Symfony/Component/Translation/Extractor/PhpExtractor.php]: Sorry, for some reason I thought `parseTokens` was `private`. Then it would not be a BC break, right? Clever workaround! I'll put it in.

File: src/Symfony/Component/Translation/Tests/Extractor/PhpExtractorTest.php [hunk 2]
```diff
@@ -68,11 +68,15 @@ public function testExtraction($resource)
...
+
+        $filename = implode(\DIRECTORY_SEPARATOR, [__DIR__, '..', 'fixtures', 'extractor', 'translation.html.php']);
+        $this->assertEquals(['sources' => [$filename.':2']], $catalogue->getMetadata('single-quoted key'));
+        $this->assertEquals(['sources' => [$filename.':43']], $catalogue->getMetadata('other-domain-test-no-params-short-array', 'not_messages'));
     }
 
     public function resourcesProvider()
     {
-        $directory = __DIR__.'/../fixtures/extractor/';
+        $directory = implode(\DIRECTORY_SEPARATOR, [__DIR__, '..', 'fixtures', 'extractor']);
```
Comment (fabpot) [inline src/Symfony/Component/Translation/Tests/Extractor/PhpExtractorTest.php]: Is this change really needed?

File: src/Symfony/Component/Translation/Tests/Extractor/PhpExtractorTest.php [hunk 3]
```diff
@@ -87,9 +91,9 @@ public function resourcesProvider()
         return [
             [$directory],
             [$phpFile],
-            [glob($directory.'*')],
+            [glob($directory.\DIRECTORY_SEPARATOR.'*')],
```
Comment (fabpot) [inline src/Symfony/Component/Translation/Tests/Extractor/PhpExtractorTest.php]: It should work with "just" `/`, no? I mean, even on Windows.

**Review by fabpot — CHANGES_REQUESTED** [review]

**Review by fabpot — APPROVED** [review]

File: src/Symfony/Component/Translation/Tests/Extractor/PhpExtractorTest.php [hunk 4]
```diff
@@ -68,11 +68,15 @@ public function testExtraction($resource)
...
+
+        $filename = __DIR__.'/../fixtures/extractor/translation.html.php';
+        $this->assertEquals(['sources' => [$filename.':2']], $catalogue->getMetadata('single-quoted key'));
+        $this->assertEquals(['sources' => [$filename.':43']], $catalogue->getMetadata('other-domain-test-no-params-short-array', 'not_messages'));
     }
 
     public function resourcesProvider()
     {
-        $directory = __DIR__.'/../fixtures/extractor/';
+        $directory = __DIR__.'/../fixtures/extractor';
```
Comment (fabpot) [inline src/Symfony/Component/Translation/Tests/Extractor/PhpExtractorTest.php]: I would revert this change as well to ease merging older branches to newer ones.

Comment (Stadly) [reply src/Symfony/Component/Translation/Tests/Extractor/PhpExtractorTest.php]: This change is needed due tests failing on Travis with `PHP 7.3` and `deps=low`: https://travis-ci.org/symfony/symfony/builds/524492376

**Comment by Stadly** [comment]
I don't think the failing build has anything to do with this PR.

File: src/Symfony/Component/Translation/Tests/Extractor/PhpExtractorTest.php (hunk 3 above)
Comment (Stadly) [reply src/Symfony/Component/Translation/Tests/Extractor/PhpExtractorTest.php]: The functionality works with `/`, but the comparisons on line `73` and `74` do not.

File: src/Symfony/Component/Translation/Tests/Extractor/PhpExtractorTest.php (hunk 2 above)
Comment (Stadly) [reply src/Symfony/Component/Translation/Tests/Extractor/PhpExtractorTest.php]: You're right. Only the trailing directory separator needs to be removed. Fixed.

File: src/Symfony/Component/Translation/Tests/Extractor/PhpExtractorTest.php [hunk 5]
```diff
@@ -69,11 +69,15 @@ public function testExtraction($resource)
         $actualCatalogue = $catalogue->all();
 
         $this->assertEquals($expectedCatalogue, $actualCatalogue);
+
+        $filename = __DIR__.'/../fixtures/extractor'.\DIRECTORY_SEPARATOR.'translation.html.php';
```
Comment (fabpot) [inline src/Symfony/Component/Translation/Tests/Extractor/PhpExtractorTest.php]: IIUC, the filename might mix both `/` and `\\` (on Windows at least). Can we standardize to always use `/` instead?

**Review by fabpot — CHANGES_REQUESTED** [review]

**Comment by fabpot** [comment]
@Stadly I see that yo merge a branch here. Can you rebase to get rid of the merge commit?

File: src/Symfony/Component/Translation/Extractor/PhpExtractor.php [hunk 6]
```diff
@@ -236,6 +242,10 @@ protected function parseTokens($tokens, MessageCatalogue $catalog)
 
                 if ($message) {
                     $catalog->set($message, $this->prefix.$message, $domain);
+                    $metadata = $catalog->getMetadata($message, $domain) ?? [];
+                    $normalizedFilename = preg_replace('{[\\\\/]+}', '/', $filename);
```
Comment (fabpot) [inline src/Symfony/Component/Translation/Extractor/PhpExtractor.php]: Can't we use `str_replace('\\', '/', $filename)` instead?

File: src/Symfony/Component/Translation/Tests/Extractor/PhpExtractorTest.php [hunk 7]
```diff
@@ -88,9 +92,9 @@ public function resourcesProvider()
         return [
             [$directory],
             [$phpFile],
-            [glob($directory.'*')],
+            [glob($directory.\DIRECTORY_SEPARATOR.'*')],
```
Comment (fabpot) [inline src/Symfony/Component/Translation/Tests/Extractor/PhpExtractorTest.php]: I suppose that this change is not needed anymore, right?

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Translation/Tests/Extractor/PhpExtractorTest.php (hunk 5 above)
Comment (Stadly) [reply src/Symfony/Component/Translation/Tests/Extractor/PhpExtractorTest.php]: Yes. Good idea!

File: src/Symfony/Component/Translation/Tests/Extractor/PhpExtractorTest.php (hunk 7 above)
Comment (Stadly) [reply src/Symfony/Component/Translation/Tests/Extractor/PhpExtractorTest.php]: Correct.

File: src/Symfony/Component/Translation/Extractor/PhpExtractor.php (hunk 6 above)
Comment (Stadly) [reply src/Symfony/Component/Translation/Extractor/PhpExtractor.php]: I also replace consecutive directory separators in order to solve the issue on line 4021: https://travis-ci.org/symfony/symfony/jobs/524492379 without changing `resourcesProvider()`. Alternatively we can use `str_replace('\\', '/', $filename)` and reintroduce the changes to `resourcesProvider()` (remove trailing `/` of `$directory` and add it to the `glob()` calls).

**Comment by Stadly** [comment]
Haha. I'm messing up everything. I'll try to fix it :)

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
One last question about this PR: we set the filename as an absolute path. Is it useful? Wouldn't it be "better" to use a relative path? I'm not sure what the spec says about this.

**Comment by Stadly** [comment]
The path used when extracting is the one set as source, so whether it is absolute or relative depends on the user. `extract(__DIR__.'/../src/')` will give absolute paths, while `extract('../src/')` will give relative paths.

**Comment by Stadly** [comment]
Should I add a test using relative paths?

**Comment by fabpot** [comment]
Thank you @Stadly.
